cpp/src/cylon/indexing/index.hpp

```diff
@@ -0,0 +1,174 @@
+#ifndef CYLON_SRC_CYLON_INDEXING_INDEX_H_
+#define CYLON_SRC_CYLON_INDEXING_INDEX_H_
+
+#include <arrow/table.h>
+#include <arrow/api.h>
+#include <arrow/compute/api.h>
+#include <arrow/compute/kernel.h>
+#include <arrow/arrow_comparator.hpp>
+#include "index.hpp"
+#include "util/macros.hpp"
+#include "status.hpp"
+#include "ctx/cylon_context.hpp"
+#include "ctx/arrow_memory_pool_utils.hpp"
+
+namespace cylon {
+
+class BaseIndex {
+
+ public:
+  explicit BaseIndex(int col_id, int size) {
+    col_id_ = col_id;
+    size_ = size;
+  };
+
+  // TODO: virtual destructor
+  //virtual void SetIndex(void *index_object) = 0;
+  //virtual void *GetIndexSet() = 0;
+
+  virtual Status Find(void *search_param,
+                    std::shared_ptr<cylon::CylonContext> &ctx,
+                    std::shared_ptr<arrow::Table> &input,
+                    std::shared_ptr<arrow::Table> &output) = 0;
+
+  virtual std::shared_ptr<arrow::Array> GetIndex() = 0;
+
+  int GetColId() const {
+    return col_id_;
+  }
+  int GetSize() const {
+    return size_;
+  }
+ private:
+  int size_;
+  int col_id_;
+};
+
+template<class ARROW_T, typename CTYPE = typename ARROW_T::c_type>
+class Index : public BaseIndex {
+ public:
+  using ARROW_ARRAY_TYPE = typename arrow::TypeTraits<ARROW_T>::ArrayType;
+  using MMAP_TYPE = typename std::unordered_multimap<CTYPE, int64_t>;
+  explicit Index(int col_ids, int size, MMAP_TYPE map) : BaseIndex(col_ids, size) {
+    map_ = map;
+  };
+
+  Status Find(void *search_param,
+              std::shared_ptr<cylon::CylonContext> &ctx,
```

Comment:
shall we accept an arrow::MemoryPool here. I think there's no need to get the ctx